Restructure content tab columns

zzandland opened the pull request:
#7348 addresses an issue where `size` column displays total size of all files in the folder, not only the selected files. This PR provides a quick fix to update folder's size from `recalculateProgress` and `recalculateAvailability`. I intentionally left out `increaseSize` function (originally used to calculate total size) in case we want to add a new column for `total size`.

glassez commented:
I recall that someone has already tried to do the same thing, but was rejected.
@qbittorrent/bug-handlers, ping.

xavier2k6 commented:
There were 3 solutions discussed in #7348 

- Rename `Size Column` to `Total Size Column` _(Easiest & avoids adding complexity)_
- Add an actual `Size Column` (**Rejected**- unless someone submitted a `Feature Request`)
- update the size for folders in the format `X / Total GiB` if files belonging to them are unselected _(This PR based on this concept)_

zzandland commented:
> update the size for folders in the format X / Total GiB if files belonging to them are unselected (This PR based on this concept)
Oops forgot to implement this.

zzandland commented:
<img width="568" alt="Untitled" src="https://user-images.githubusercontent.com/26905303/107688264-4d9f2b00-6c5c-11eb-87e5-6740de682407.png">

Decided to leave units for selected size as unit can differ between it and total size.

thalieht commented:
Wouldn't it be better to show the "X /" part only when selectedSize < totalSize?

glassez commented:
Personally, I haven't changed my mind since last time. If we have a Size column that displays the size for the files, then we expect the folders to also display their size (i.e., the sum of the sizes of all the files/folders inside).
Why don't you just add a special column for "download size", which will be 0 for unchecked files and the sum of sub-item "download sizes" for folders?
Aggregated column looks bad, IMO.

xavier2k6 commented:
>Why don't you just add a special column for "download size", which will be 0 for unchecked files and the sum of sub-item "download sizes" for folders?

That's effectively this, isn't it?? (Rename `Size Column` to `Total Size Column` & Add an actual `Size Column`)

If you observe the `Transfer List` with these two columns displayed - the `Size` column will show same or similar when only selected files are chosen to download!

zzandland commented:
> Personally, I haven't changed my mind since last time. If we have a Size column that displays the size for the files, then we expect the folders to also display their size (i.e., the sum of the sizes of all the files/folders inside).

I think the confusion stems from the fact that definition of "size" is arbitrary. It can mean the *absolute* size of the folder regardless of what is downloaded, but also the *relative* size of the folder which is the sum of downloading files.

> Aggregated column looks bad, IMO.

Yeah I also think the same. 

> Why don't you just add a special column for "download size", which will be 0 for unchecked files and the sum of sub-item "download sizes" for folders?

What's the point of the `remaining` column then? On hindsight, "download size" is just the reverse of the "remaining".

How about we just rename the "size" column to "total size" to have a clear meaning, then change the position of the "remaining" column left of the "total size" column so that users can focus more on the size what they're downloading?

glassez commented:
>How about we just rename the "size" column to "total size" to have a clear meaning

Clear meaning??? It looks like you're trying to confuse me. The clear meaning is when the Size of file means exactly "file size". The clear meaning is when the Size of a folder means the sum of the sizes of the files it contains. 
>What's the point of the remaining column then? On hindsight, "download size" is just the reverse of the "remaining".

By "download size" I meant the size I want to download from this file/folder.

The `Total size` of file looks silly to me if I see just the `Size` of file next to it, since the Size means the same by default. It would be acceptable if all other Sizes were used with the defining word. But then we come back to the question, why not use Size in its basic sense without additional definitions?

zzandland commented:
> The Total size of file looks silly to me if I see just the Size of file next to it, since the Size means the same by default. It would be acceptable if all other Sizes were used with the defining word. But then we come back to the question, why not use Size in its basic sense without additional definitions?

Yeah that would look quite weird if I were to see 'total size' right next to 'size'. It would make more sense to see 'size' and 'download size'. Should I update this PR in this direction then?

Chocobo1 commented:
> > The Total size of file looks silly to me if I see just the Size of file next to it, since the Size means the same by default. It would be acceptable if all other Sizes were used with the defining word. But then we come back to the question, why not use Size in its basic sense without additional definitions?
> 
> Yeah that would look quite weird if I were to see 'total size' right next to 'size'. It would make more sense to see 'size' and 'download size'. Should I update this PR in this direction then?

Probably worth mentioning that utorrent uses "Size" as the total size of a file and "Done" as the downloaded size. I find it rather intuitive to understand what it means.

glassez commented:
> Probably worth mentioning that utorrent uses "Size" as the total size of a file and "Done" as the downloaded size. I find it rather intuitive to understand what it means.

As far as I understood, it is not about the size already downloaded, but about the size that the user wants to download (i.e. it is unchanged during downloading and affected only by selecting/unselecting files). So `Done` doesn't fit.

Chocobo1 commented:
>As far as I understood, it is not about the size already downloaded, but about the size that the user wants to download (i.e. it is unchanged during downloading and affected only by selecting/unselecting files). So Done doesn't fit.

OK, forget about "Done".

>If we have a Size column that displays the size for the files, then we expect the folders to also display their size (i.e., the sum of the sizes of all the files/folders inside).

I agree with the part of summing the file sizes within a folder.

>Why don't you just add a special column for "download size", which will be 0 for unchecked files and the sum of sub-item "download sizes" for folders?

We already have "Remaining" column for this.

I'm not sure if users requested this and I won't mind we have another "Total size" column which folder size won't depend on files selected or not and file size also won't depend on selection.

Chocobo1 commented:
To recap, I imagine we would define the following columns:

|  | Size | Total size | Done | Progress (%) | Remaining |
| ---      | ---  |  --- | --- | --- | --- |
| File | File size (`0` if unselected) | File size (won't depend on selection) | Size downloaed | `Done / Total_size` | if unselected `0` otherwise `Total size - Done` |
| Folder | Folder size (depends on files' selection) | Folder size (won't depend on selection) | Size downloaed  | `Done / Total_size` | if unselected `0` otherwise `Total size - Done`  |

glassez commented:
>Progress (%) = Done / Total size

Are you sure? It will never reach 100 % for folders where some of the files are unselected.
>It might be a good idea to rename "Size" -> "Download size" and "Total size" -> "Size".

That's exactly what I'm talking about. The Size of file has a stable meaning so it isn't good idea to use in a different sense. So I still suggest using Size in the sense of "full size" of the file/folder, and in all other senses use the defining word. Otherwise we can use  defining word with all sizes then "Total size" will be appropriate. BTW, the word Size can be omitted for compactness so we can end up with the following column names: Total, Wanted, Done, Remaining, etc.).

Chocobo1 commented:
>Are you sure? It will never reach 100 % for folders where some of the files are unselected.

Updated, it's not easy to get it right on the first try.

ArcticGems commented:
> 
> It might be a good idea to rename "Size" -> "Download size" 

Maybe even shorter = "DL size" (good for small screens like smartphones/tablets).

zzandland commented:
> To recap, I imagine we would define the following columns:
> 
> Size	Total size	Done	Progress (%)	Remaining
> File	File size (`0` if unselected)	File size (won't depend on selection)	Size downloaed	`Done / Total size`	If unselected `0` otherwise `Total size - Done`
> Folder	Folder size (depends on files' selection)	Folder size (won't depend on selection)	Size downloaed	`Done / Size`	If unselected `0` otherwise `Total size - Done`
> It might be a good idea to rename "Size" -> "Download size" and "Total size" -> "Size".

I think this is a pretty good implementation goal. I'll proceed with changing the PR to reflect these changes.

zzandland commented:
![image](https://user-images.githubusercontent.com/26905303/108139331-be877e00-7074-11eb-9e7b-122744da42b0.png)

Added DL Size and Done column.

zzandland commented:
<img width="672" alt="Untitled" src="https://user-images.githubusercontent.com/26905303/108139550-23db6f00-7075-11eb-9647-9450baff629c.png">

Added `DL Size` and `Done` columns. Moved `Download Priority` column to be on the furthest right.

zzandland commented on src/gui/torrentcontentmodelfile.cpp:
```diff
@@ -29,6 +29,7 @@
 #include "torrentcontentmodelfile.h"
 
 #include "base/bittorrent/common.h"
+#include "base/bittorrent/downloadpriority.h"
```
Shouldn't be here. I will delete it.

zzandland commented on src/gui/torrentcontentmodelfile.cpp:
```diff
@@ -59,11 +60,19 @@ void TorrentContentModelFile::setPriority(BitTorrent::DownloadPriority newPriori
     if (m_priority == newPriority)
         return;
 
+    if (m_priority != BitTorrent::DownloadPriority::Ignored && newPriority == BitTorrent::DownloadPriority::Ignored)
+      m_parentItem->updateDLSize(-m_size);
+    else if (m_priority == BitTorrent::DownloadPriority::Ignored && newPriority != BitTorrent::DownloadPriority::Ignored)
+      m_parentItem->updateDLSize(m_size);
```
DL Size of all upper directories should be updated in bottom-up fashion.

zzandland commented on src/gui/torrentcontentmodelfile.cpp:
```diff
@@ -59,11 +60,19 @@ void TorrentContentModelFile::setPriority(BitTorrent::DownloadPriority newPriori
 
     // Update parent
     if (updateParent)
+    {
         m_parentItem->updatePriority();
+    }
```
Will remove bracket

thalieht commented:
This is what i would think as a user, if i didn't know anything about how it works:
DL size and Size both show some size for files which, upon further inspection, happen to be the same. I would continue to monitor these columns waiting for a file to have different DL size and Size because why else would it show it twice right? Perhaps i would never realize that these 2 columns will always be the same for selected files but if i did, my first thought would be "what a poor UI choice... at least now that i know, i can always mentally ignore DL size for files and focus only on folders". 

Just my 2 cents and no i don't have any other ideas.

zzandland commented:
@thalieht when the user clicks the checkbox off for a file, they will see that the DL size turns to 0. Wouldn't they figure out then?

glassez commented:
@thalieht, maybe column names proposed in https://github.com/qbittorrent/qBittorrent/pull/14389#issuecomment-778632269 are more clear? Total, Wanted, Done, Remaining.

zzandland commented:
@glassez I didn't see that comment! That sounds more clear to me.

ArcticGems commented:
Maybe "Selected" instead of "Wanted"??

thalieht commented:
Column names are irrelevant for this. The point is that a whole new column is introduced just for the folders. For files we can just look at the file checkboxes and it serves the exact same purpose as this faux binary representation (0 or file size) of "DL size" (name doesn't matter).

glassez commented:
I'm sorry, but then I don't know what would suit you... The inconsistent/confusing representation, when in the same column we have different information for files (their size) and for folders (full size/size of selected files)? Or an even more illogical representation, when for folders we only show the size of the selected files in the Size column, so that its value does not equal the sum of the nested values?

zzandland commented:
<img width="696" alt="Untitled" src="https://user-images.githubusercontent.com/26905303/108243216-b413c580-7102-11eb-9dcb-6403fcbd6ea1.png">

Updated version. I moved progress and download priority columns to be next of the name column in a separate commit. I will revert it if there's too much resistance.

zzandland commented:
> Column names are irrelevant for this. The point is that a whole new column is introduced just for the folders. For files we can just look at the file checkboxes and it serves the exact same purpose as this faux binary representation (0 or file size) of "DL size" (name doesn't matter).

@thalieht I agree that this PR adds a redundant column for files, but we do need to provide wanted size for folders. I believe that this is a fair tradeoff.

glassez commented:
> I'm sorry, but then I don't know what would suit you... The inconsistent/confusing representation, when in the same column we have different information for files (their size) and for folders (full size/size of selected files)? Or an even more illogical representation, when for folders we only show the size of the selected files in the Size column, so that its value does not equal the sum of the nested values?

Although I wouldn't be surprised if what seems illogical to me turns out to be exactly what most people want.

ArcticGems commented:
> 
> 
> <img alt="Untitled" width="696" src="https://user-images.githubusercontent.com/26905303/108243216-b413c580-7102-11eb-9dcb-6403fcbd6ea1.png">
> 
> Updated version. I moved progress and download priority columns to be next of the name column in a separate commit. I will revert it if there's too much resistance.

I personally would want it in only **one** column called "Wanted/Total"(example "347MiB/1.46 GiB) and maybe have a toggle option to hide "Total" size in Tools/Options for those users who wants it.

thalieht commented:
>I'm sorry, but then I don't know what would suit you..

Well, i don't know what else to say. I don't feel strongly about any of this, i just expressed my opinion. For the record i see nothing wrong with https://github.com/qbittorrent/qBittorrent/pull/14389#issuecomment-777733876. Maybe not the best solution but definitely better than the new column thing IMO.

glassez commented:
>For the record i see nothing wrong with [#14389 (comment)](https://github.com/qbittorrent/qBittorrent/pull/14389#issuecomment-777733876). Maybe not the best solution but definitely better than the new column thing IMO.

Well, if most people prefer this option, I won't mind.

ArcticGems commented:
> 
> 
> For the record i see nothing wrong with [#14389 (comment)](https://github.com/qbittorrent/qBittorrent/pull/14389#issuecomment-777733876). Maybe not the best solution but definitely better than the new column thing IMO.

Definitely the best of solution ln this PR! :+1: 

Skip the new column idea.

xavier2k6 commented:
>  For the record i see nothing wrong with [#14389 (comment)](https://github.com/qbittorrent/qBittorrent/pull/14389#issuecomment-777733876). Maybe not the best solution but definitely better than the new column thing IMO.

I would suggest renaming the column though from `Size` to `Selected/Original` or `Wanted/Original` (Omit the Size)

zzandland commented:
To be honest I like the 4 column idea (wanted, total, done, remaining) because it’s a coherent way of displaying all size information. If I were to revert back to a single column idea I think column name shouldn’t change but stick to ‘size’ because showing as ‘selected/total’ would only happen for folders with partially selected files. Otherwise even for unselected files I need to show as 0/53.3kB for column name ‘selected/total’ to make sense.

glassez commented:
> If I were to revert back to a single column idea I think column name shouldn’t change but stick to ‘size’

👍

Chocobo1 commented:
>To be honest I like the 4 column idea (wanted, total, done, remaining) because it’s a coherent way of displaying all size information.

If it is possible to hide columns and by default show only one of the size columns (the one which most users expect), then I suppose it would meet all expectations.

ArcticGems commented:
> 
> 
> > To be honest I like the 4 column idea (wanted, total, done, remaining) because it’s a coherent way of displaying all size information.
> 
> If it is possible to hide columns and by default show only one of the size columns (the one which most users expect), then I suppose it would meet all expectations.

Wanted and Total can't be sorted differently from each other (right?), so makes no sense to me to have 2 columns.

Just like in https://github.com/qbittorrent/qBittorrent/pull/14389#issuecomment-777733876 , you should only show the 2 sizes for folders. Users see a partially checked folder then they see for example "78.0 KiB/544.9MiB". It's simple and logical. :+1:

sakkamade commented:
> It's simple and logical. 👍

However is utterly distinct from what `qbit` user used, and expect, to see, 
for there was yet none such one-column for two different sizes values.

As for me, these four colums appearing to be a right choice. 
Although ability to hide some of the columns (and the `Progress` is amongst them)  would be great to have, indeed.

zzandland commented:
> As for me, these four colums appearing to be a right choice.
> Although ability to hide some of the columns (and the Progress is amongst them) would be great to have, indeed.

👍 

> Edit:
> As for column names I'd rather chose Size (instead of Wanted it is), Total Size, Remaining, Completed — names from the main window.

Remaining and Completed (or Done) all imply size. Wouldn't it be better if the other two also imply `size` but follow the same wording style? Thus Selected and Total.

sakkamade commented:
> Wouldn't it be better if the other two also imply size but follow the same wording style? Thus Selected and Total.

I believe since the `Size` and `Total Size` are father coupled, they should have exaclty such names which are duly serves an indication of it. 
For the same reason `Remaining` and `Completed` appearing to me the finest pick.

zzandland commented:
I'd like to know how everyone wants to proceed with this PR. I prefer having 4 columns with Selected, Total, Remaining, Completed, but if the majority would prefer having 3 columns as Size(`Selected/Total`), Remaining, Completed then I can make changes to reflect it. Please leave your preference as a comment (3 columns or 4 columns)

ArcticGems commented:
I vote for **3** columns (hopefully something like https://github.com/qbittorrent/qBittorrent/pull/14389#issuecomment-782004108)

(Upvote this comment or create a comment about the  **4** columns idea and have people upvote it - this will help zzandland I think.)
